check.go

```diff
@@ -0,0 +1,1086 @@
+							referencedLayers[layer] = true
+						}
+						if _, ok := referencedROLayers[layer]; ok {
+							referencedROLayers[layer] = true
+						}
+					}
+					if isReadWrite {
+						if len(report.Layers[layer]) > 0 {
+							report.Images[id] = append(report.Images[id], report.Layers[layer]...)
+						}
+						if len(report.ROLayers[layer]) > 0 {
+							report.Images[id] = append(report.Images[id], report.ROLayers[layer]...)
+						}
+					} else {
+						if len(report.Layers[layer]) > 0 {
+							report.ROImages[id] = append(report.ROImages[id], report.Layers[layer]...)
+						}
+						if len(report.ROLayers[layer]) > 0 {
+							report.ROImages[id] = append(report.ROImages[id], report.ROLayers[layer]...)
+						}
+					}
+				}
+			}
+		}
+		return false, nil
+	}); errored {
+		return CheckReport{}, err
+	}
+
+	// Iterate over each container in turn.
+	if done, err := s.readContainerStore(func() (bool, error) {
+		containers, err := s.containerStore.Containers()
+		if err != nil {
+			return true, err
+		}
+		for i := range containers {
+			container := containers[i]
+			id := container.ID
+			logrus.Debugf("checking container %s", id)
+			if options.ContainerData {
+				// Check that all of the big data items are present and reading them
+				// back gives us the right amount of data.
+				for _, key := range container.BigDataNames {
+					func() {
+						data, err := s.containerStore.BigData(id, key)
+						if err != nil {
+							if errors.Is(err, os.ErrNotExist) {
+								err = fmt.Errorf("container %s: data item %q: %w", id, key, ErrContainerDataMissing)
+								report.Containers[id] = append(report.Containers[id], err)
+								return
+							}
+							err = fmt.Errorf("container %s: data item %q: %w", id, key, err)
+							report.Containers[id] = append(report.Containers[id], err)
+							return
+						}
+						if int64(len(data)) != container.BigDataSizes[key] {
+							err = fmt.Errorf("container %s: data item %q: %w", id, key, ErrContainerDataIncorrectSize)
+							report.Containers[id] = append(report.Containers[id], err)
+							return
+						}
+					}()
+				}
+			}
+			// Look at the container's base image.  If the image has errors, the image's errors
+			// are the container's errors.
+			if container.ImageID != "" {
+				if _, checked := examinedImages[container.ImageID]; !checked {
+					err := fmt.Errorf("image %s: %w", container.ImageID, ErrContainerImageMissing)
+					report.Containers[id] = append(report.Containers[id], err)
+				}
+				if len(report.Images[container.ImageID]) > 0 {
+					report.Containers[id] = append(report.Containers[id], report.Images[container.ImageID]...)
+				}
+				if len(report.ROImages[container.ImageID]) > 0 {
+					report.Containers[id] = append(report.Containers[id], report.ROImages[container.ImageID]...)
+				}
+			}
+			// Count the container's layer as referenced.
+			if container.LayerID != "" {
+				referencedLayers[container.LayerID] = true
```

Comment:
What if this layer reference is dangling? What if this points to a read-only layer?